# **Dataffirm - Technical Test**

---

# Contents
1. [Intro](#Intro)   
  
2. [Data Exploration](#Data-Exploration)  
  
3. [Data Cleaning](#Data-Cleaning)
    - [Missing `Track_Name`](#Missing-Track_Name)
    - [Missing `Artist_ID`](#Missing-Artist_ID)
    - [Missing `Track_ID`](#Missing-Track_ID)
    - [Alternative Approach](#Alternative-Approach)  
  
4. [Data Analysis](#Data-Analysis)
    - [A.](#A.)
    - [B.](#B.)
    - [C.](#C.)
    
---

# Intro

This is my solution to the technical test put forward by Dataffirm. The data recieved was file of a 2.53 GB in size, in a file called:
`userid-timestamp-artid-artname-traid-traname`

Assuming that the name of this file is the name of the columns, this gives us:
- User ID
- Timestamp
- Artist ID
- Artist Name
- Track ID
- Track Name

These columns will now be referred to the above respectively, with all spaces replaced with "\_" i.e. `User_ID`.

---

# Data Exploration

In [10]:
import pandas as pd

pd.set_option('display.max_rows', 100)

header_list = ["User_ID", "Timestamp", "Artist_ID", "Artist_Name", "Track_ID", "Track_Name"]

data = 'data.tsv'

df = pd.read_csv(data, sep='\t', header=None, names=header_list)

>Imported the base pandas library which I'll be using for my task today of Exploration, Cleansing and Analysis. I have also created a list of the headers I want for the data that will be loaded and the file location of the data to load.
---

In [11]:
df.shape

(19098862, 6)

_fig 2.1_

---

In [3]:
df.head()

User_ID             Timestamp                             Artist_ID  \
0  user_000001  2009-05-04T23:08:57Z  f1b1cf71-bd35-4e99-8624-24a6e15f133a   
1  user_000001  2009-05-04T13:54:10Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
2  user_000001  2009-05-04T13:52:04Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
3  user_000001  2009-05-04T13:42:52Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
4  user_000001  2009-05-04T13:42:11Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   

  Artist_Name Track_ID                                  Track_Name  
0   Deep Dish      NaN  Fuck Me Im Famous (Pacha Ibiza)-09-28-2007  
1        坂本龍一      NaN           Composition 0919 (Live_2009_4_15)  
2        坂本龍一      NaN                        Mc2 (Live_2009_4_15)  
3        坂本龍一      NaN                     Hibari (Live_2009_4_15)  
4        坂本龍一      NaN                        Mc1 (Live_2009_4_15)

_fig 2.2_

---

In [4]:
df.describe()

User_ID             Timestamp  \
count      19098862              19098862   
unique          992              17454739   
top     user_000949  2009-02-26T21:29:15Z   
freq         183103                   248   

                                   Artist_ID Artist_Name  \
count                               18498009    19098862   
unique                                107295      173923   
top     a74b1b7f-71a5-4011-9441-d0b5e4122711   Radiohead   
freq                                  115099      115099   

                                    Track_ID Track_Name  
count                               16936136   19098850  
unique                                960402    1083480  
top     db16d0b3-b8ce-4aa8-a11a-e4d53cc7f8a6      Intro  
freq                                    3991      17561

_fig 2.3_

---

In [5]:
df.isna().sum()

User_ID              0
Timestamp            0
Artist_ID       600853
Artist_Name          0
Track_ID       2162726
Track_Name          12
dtype: int64

_fig 2.4_

---

In [6]:
minDate = df['Timestamp'].min()
maxDate = df['Timestamp'].max()
print('Min. date = {}\nMax. date = {}'.format(minDate, maxDate))

Min. date = 2005-02-14T00:00:07Z
Max. date = 2013-09-29T18:32:04Z


_fig 2.5_

---

> From the above (_fig 2.1 - 2.5_) it can be seen that there are 19098862 values in the dataset across the 6 columns I mentioned above. It is panel data, defining the tracks and their respective artists that users have listened to between 14/02/2005 - 29/09/2013.
>
> There is missing data in 3 of the categories, culminating to 600853, 2162726, 12 cases for `Artist_ID`, `Track_ID` and `Track_Name` respectively. These will have to be mitigated in the next step, data cleaning.

# Data Cleaning
>Given the missing values in the dataset, I will start by trying to recover as much data as possible, so as to try to enrich the data in the format I was originally provided it in.

## Missing `Track_Name`
>There are 12 records that are missing in the `Track_Name` field, as shown in the [data exploration](#Data-Exploration) above.

>I am hoping to find that these records have `Track_ID`s despite not having any `Track_Name`s, as this will allow me to identify other records with the same `Track_ID` and therefore find the associated `Track_Name` for the record.

In [7]:
df.loc[df['Track_Name'].isna()]

User_ID             Timestamp  \
1264044   user_000060  2006-04-19T10:41:44Z   
1371571   user_000067  2006-06-08T13:17:00Z   
1371606   user_000067  2006-06-07T18:25:19Z   
1371738   user_000067  2006-06-06T14:09:10Z   
1371955   user_000067  2006-05-30T14:02:33Z   
1372180   user_000067  2006-05-25T16:14:33Z   
1372494   user_000067  2006-05-15T20:20:25Z   
1372970   user_000067  2006-04-13T19:47:12Z   
1373492   user_000067  2006-03-21T16:41:56Z   
1373526   user_000067  2006-03-20T21:08:07Z   
1374009   user_000067  2006-03-08T19:24:07Z   
10975675  user_000585  2009-01-14T03:56:02Z   

                                     Artist_ID  \
1264044   1f4bf518-3aa3-4114-ada7-64884e580ad2   
1371571   945c0e49-8d06-4f01-87cf-c2d82b00d5ac   
1371606   945c0e49-8d06-4f01-87cf-c2d82b00d5ac   
1371738   945c0e49-8d06-4f01-87cf-c2d82b00d5ac   
1371955   945c0e49-8d06-4f01-87cf-c2d82b00d5ac   
1372180   1f4bf518-3aa3-4114-ada7-64884e580ad2   
1372494   945c0e49-8d06-4f01-87cf-c2d82b00d5ac   
1372970   1f4bf518-3aa3-4114-ada7-64884e580ad2   
1373492   945c0e49-8d06-4f01-87cf-c2d82b00d5ac   
1373526   1f4bf518-3aa3-4114-ada7-64884e580ad2   
1374009                                    NaN   
10975675                                   NaN   

                                                Artist_Name Track_ID  \
1264044                                         The Sammies      NaN   
1371571                                         Simon Dawes      NaN   
1371606                                         Simon Dawes      NaN   
1371738                                         Simon Dawes      NaN   
1371955                                         Simon Dawes      NaN   
1372180                                         The Sammies      NaN   
1372494                                         Simon Dawes      NaN   
1372970                                         The Sammies      NaN   
1373492                                         Simon Dawes      NaN   
1373526                                         The Sammies      NaN   
1374009                       Go Gettas Entertainment Feat.      NaN   
10975675  Silence'' Foundation Laboratory (Silentpoets)\...      NaN   

         Track_Name  
1264044         NaN  
1371571         NaN  
1371606         NaN  
1371738         NaN  
1371955         NaN  
1372180         NaN  
1372494         NaN  
1372970         NaN  
1373492         NaN  
1373526         NaN  
1374009         NaN  
10975675        NaN

_fig 3.1_

---

>As can be seen above (_fig 3.1_), the records with missing `Track_Name`s are also missing any associated `Track_ID`. This means that I will be unable to identify these tracks individually, unless I am able to assume that these Artists (`The Sammies`, `Simon Dawes`, `Go Gettas Entertainment Feat.`,
`Silence'' Foundation Laboratory (Silentpoets)\...`) have only a single track ever released.

In [8]:
df.loc[df['Artist_Name'] == 'The Sammies']

User_ID             Timestamp  \
1264044   user_000060  2006-04-19T10:41:44Z   
1334134   user_000067  2009-04-06T19:59:06Z   
1372180   user_000067  2006-05-25T16:14:33Z   
1372970   user_000067  2006-04-13T19:47:12Z   
1373526   user_000067  2006-03-20T21:08:07Z   
17287587  user_000896  2009-05-03T23:09:58Z   
17287588  user_000896  2009-05-03T23:06:41Z   
17287589  user_000896  2009-05-03T23:03:41Z   
17287590  user_000896  2009-05-03T23:00:45Z   
17287591  user_000896  2009-05-03T22:57:02Z   
17287592  user_000896  2009-05-03T22:53:35Z   
17287593  user_000896  2009-05-03T22:50:36Z   
17287594  user_000896  2009-05-03T22:47:47Z   
17287595  user_000896  2009-05-03T22:44:41Z   
17287596  user_000896  2009-05-03T22:41:43Z   
17287597  user_000896  2009-05-03T22:38:58Z   
17287598  user_000896  2009-05-03T22:25:03Z   
17287599  user_000896  2009-05-03T22:21:32Z   
17298059  user_000896  2007-05-08T05:37:44Z   
17298063  user_000896  2007-05-07T22:51:26Z   
17298064  user_000896  2007-05-07T22:23:19Z   
17298450  user_000896  2007-04-10T22:31:46Z   
17298451  user_000896  2007-04-10T22:28:30Z   
17298452  user_000896  2007-04-10T22:26:02Z   
17298453  user_000896  2007-04-10T22:23:32Z   
17298454  user_000896  2007-04-10T22:21:12Z   
17298455  user_000896  2007-04-10T22:18:16Z   
17298456  user_000896  2007-04-10T22:13:44Z   
17298457  user_000896  2007-04-10T22:10:38Z   
17298458  user_000896  2007-04-10T22:07:36Z   
17298459  user_000896  2007-04-10T22:01:41Z   
17298460  user_000896  2007-04-10T21:57:37Z   
17298461  user_000896  2007-04-10T21:53:09Z   
17298462  user_000896  2007-04-10T21:50:06Z   

                                     Artist_ID  Artist_Name  \
1264044   1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
1334134   1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
1372180   1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
1372970   1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
1373526   1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287587  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287588  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287589  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287590  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287591  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287592  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287593  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287594  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287595  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287596  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287597  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287598  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17287599  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298059  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298063  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298064  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298450  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298451  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298452  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298453  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298454  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298455  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298456  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298457  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298458  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298459  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298460  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298461  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   
17298462  1f4bf518-3aa3-4114-ada7-64884e580ad2  The Sammies   

                                      Track_ID              Track_Name  
1264044                                    NaN                     NaN  
1334134         

_fig 3.2_

---

In [14]:
tracks_to_remove = df.loc[df['Track_Name'].isna()].index
df.drop(tracks_to_remove, inplace=True)
df.loc[df['Track_Name'].isna()]

Empty DataFrame
Columns: [User_ID, Timestamp, Artist_ID, Artist_Name, Track_ID, Track_Name]
Index: []

_fig 3.3_

---

> However, once you check for the artist across the entire dataframe (_fig 3.2_), you can quickly see that more than 1 song from any of the given bands was being listened to. Therefore, using my methodology, which assumes that the tracks that are missing `Track_ID` and `Track_Name` are the same song from any particular band, would be erroneous as there is evidence of multiple songs from the bands being listened to across the dataframe.

> This leads to the conclusion that I will not use these 12 records with missing `Track_ID` and `Track_Name` fields in my analysis, as there is no certain way for me to accomodate them.

> I'm aware that in more generalised metrics, such as "Top listened to artists", I could include these records that are only missing the finer, song-level detail. However, I believe that when extrapolated across larger datasets this would pose serious problems in terms of consistency of the data that is being used for the analysis. Due to this reason I have chosen to completely remove these 12 records and move forward with the clean dataset that is left.

> Now the dataframe no longer includes those 12 records that were missing intially (_fig 3.3_).

## Missing `Artist_ID`

>There are 600853 records that have a `Nan` value in the `Artist_ID` column. As there are no `Artist_Name`s missing, I will perform the following steps:
- List of `Artist_Name`s, whos `Artist_Name` is equal to `Nan`
- Use the aforementioned `Nan_Artist_ID_List` list to find those who have an `Artist_ID` present on another record with the same `Artist_Name`
- Use any found `Artist_ID`s to fill in the gaps of in the `Nan_Artist_ID_List`
- Check for any left over
>
>This should allow for the original data style to be left unaltered, whilst filling in any gaps that are currently present.

In [10]:
Nan_Artist_Name_List = df.loc[df['Artist_ID'].isna(), ['Artist_Name']].Artist_Name.unique().tolist()
Nan_Artist_Name_DF = df.loc[df['Artist_Name'].isin(Nan_Artist_Name_List) & df['Artist_ID'].notna(), ['Artist_Name', 'Artist_ID']]

_fig 4.1_

---

In [11]:
Found_Artist_Name = Nan_Artist_Name_DF.Artist_Name.unique().tolist()
Found_Artist_ID = Nan_Artist_Name_DF.Artist_ID.unique().tolist()
Artist_ID_dict = dict(zip(Found_Artist_Name, Found_Artist_ID))
Artist_ID_dict

{'Dёргать!': 'd800e320-49a3-44fa-af7c-058d3b337391',
 'J': '49467279-f0e9-4b55-a03b-cf66b5cf40e7',
 'Roué': '5155b18a-68d8-4260-a92f-17a8c12513a2',
 'Gülay': '863cf5d5-1667-49e4-996a-b2578108bb10',
 'Έλενα Παπαρίζου': '28a40a67-ecd3-432a-be2f-51490a7743ec',
 'ボビー・マルゴ': '66699510-3823-4137-8477-596f2b85772e',
 'Benassi Bros Feat. Dhany': 'd60e1106-6a1a-4311-a148-ae2f050a97f2',
 'にせんねんもんだい': 'b7976061-d050-4df8-ab61-2e2bcce1477f'}

_fig 4.2_

---

In [12]:
for item in Artist_ID_dict:
    df.loc[df['Artist_Name'] == item, 'Artist_ID'] = Artist_ID_dict[item]

df.loc[df['Artist_Name'] == 'にせんねんもんだい', ['Artist_Name', 'Artist_ID']]

Artist_Name                             Artist_ID
12229836   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
12229841   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
12229862   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
12229870   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
12229878   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
12229882   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
12229889   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
12229894   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
12229896   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926977   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926978   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926979   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926980   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926981   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926982   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926983   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926984   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926985   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
13926986   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f
14913426   にせんねんもんだい  b7976061-d050-4df8-ab61-2e2bcce1477f

_fig 4.3_

---

In [13]:
df.isna().sum()

User_ID              0
Timestamp            0
Artist_ID       599603
Artist_Name          0
Track_ID       2162714
Track_Name           0
dtype: int64

_fig 4.4_

---

> In _fig 4.1_ it was found that there are only 8 `Artist_Name`s that have an already existing `Artist_ID`, which isn't present in all of their records.   
This is out of 69418 unique missing `Artist_Name`s, from a total of 600851. In total these 8 bands accounted for 1250 errors throughout the dataset, which is positive but still a very marginal gain on the larger problem.

> _fig 4.2_ shows the `Artist_Name` : `Artist_ID` unique pairing that is present, with _fig 4.3_ showing the implementation of this dictionary pairing on the dataframe; including an example fo the fix.  
_fig 4.4_ shows reduction in the number of `Artist_ID` errors that was previously being experienced.



## Missing `Track_ID`

>I was going to tackle this problem in very much the same methodology I had with `Artist_ID`, as I found this method robust and easy to implement.  
However, after the only minor improvements I experienced before I decided I needed to take a different approach. This section has been typed out so you can see what I would have done if this had been more succesfull than it was in the previous stage.  
The only noticeable difference in this section from the previous is that I use the _index_ of the record instead of the _name_ field to identify where the corresponding _ID_ should be applied. This was because I saw it as being more efficient to seek out the index of the record it was editing instead of looping through the subset I was trying to fix as I had been doing earlier.  
My alternative solution was to simply re-hash all the `Artist_ID`/`Track_ID` values dependent on their respective `Artist_Name`/`Track_Name` fields. I explain why I do this in the next section [Alternative Approach](#Alternative-Approach).

In [14]:
Nan_Track_Name_List = df.loc[df['Track_ID'].isna(), ['Track_Name']].Track_Name.unique().tolist()
Nan_Track_Name_DF = df.loc[df['Track_Name'].isin(Nan_Track_Name_List) & df['Track_ID'].notna(), ['Track_Name', 'Track_ID']]

_fig 5.1_

---

In [15]:
Found_Track_Index = Nan_Track_Name_DF.index.tolist()
Found_Track_ID = Nan_Track_Name_DF.Track_ID.unique().tolist()
Track_ID_dict = dict(zip(Found_Track_Index, Found_Track_ID))
Track_ID_dict

{17: '4e78efc4-e545-47af-9617-05ff816d86e2',
 21: 'f78c95a8-9256-4757-9a9f-213df5c6854e',
 22: 'c4fc8802-d186-4c4d-85cd-d5d063b935c2',
 37: '41bacfc2-c594-4e07-a70d-3675f39003ec',
 39: '0024d72c-136f-49f2-9078-ce4b39b94d3f',
 41: '00b811a4-762b-492e-b5ef-9a673a55da57',
 43: '6b71d43e-9258-4abd-89da-921fe00070a1',
 45: '8e9ec4d9-2248-4f6f-af74-2982f107159a',
 75: 'd1de63b3-fb80-46bc-befa-8e55076b379a',
 94: 'da8e7bec-f1ce-4664-a229-1c2320ea5e05',
 109: 'afc94794-ba9f-42d3-afea-9c55f153e224',
 124: '439757fd-0309-4aba-97d6-b3163fcdab21',
 138: '6bb694b1-0d23-46e9-a03f-03a9e1628a38',
 150: 'a15efe90-1a35-4633-8888-ab4bf879527a',
 155: '402ce8e4-bbbc-4bd9-9f83-4c91d52473ad',
 158: '0e432672-675c-42fe-b164-d7fd728c0cee',
 159: 'ffb38aab-0144-4c9b-8fcf-8f6dff4362fd',
 161: 'aa25c2ca-65f0-46ff-a555-26a4e5615333',
 163: '50d1ad97-dbca-40e6-8856-f275cc4c6a6a',
 165: '977f5228-8365-411d-be4b-c26a74f01f92',
 166: 'd276b077-c05d-43c8-b168-63f12a6001b2',
 177: 'f12b2e3e-53be-47cb-aeaa-ab4b15be1ef4'

_fig 5.2_

---

In [16]:
len(Track_ID_dict)

260246

In [18]:
for item in Track_ID_dict:
    df.loc[item, 'Track_ID'] = Track_ID_dict[item]

df.loc[df['Track_Name'] == 'Elysian Fields', ['Track_Name', 'Track_ID']]

KeyboardInterrupt: 

_fig 5.3_

---

> As I stated at the [beginning of this section](#Missing-Track_ID), this section was completed purely as an example of how I would of applied the same logic for finding the already present `Artist_ID`s to `Track_ID`s.

## Alternative Approach

>With my first solution to the missing `Artist_ID`/`Track_ID` fields under performing against my expectations, I decided to re-hash all data in those respective columns.  
This was possible now because I have removed the 12 records that were lacking the `Track_Name` field that would be required for creating the `Track_ID` hash. As I stated towards the end of the ["Missing Track Name" section](#Missing-Track_Name), I removed these 12 as I believe that they would only confuse matters when it can to discussing different metrics; some of which included these 12 and others that did not. Removing them removed this potential for confusion from the discussion, which was my primary motive in this decision.

In [15]:
df['Artist_ID'] = df['Artist_Name'].apply(hash)
df['Track_ID'] = df['Track_Name'].apply(hash)
df

User_ID             Timestamp            Artist_ID  \
0         user_000001  2009-05-04T23:08:57Z  -926383222125822968   
1         user_000001  2009-05-04T13:54:10Z -3929307610419788574   
2         user_000001  2009-05-04T13:52:04Z -3929307610419788574   
3         user_000001  2009-05-04T13:42:52Z -3929307610419788574   
4         user_000001  2009-05-04T13:42:11Z -3929307610419788574   
5         user_000001  2009-05-04T13:38:31Z -3929307610419788574   
6         user_000001  2009-05-04T13:33:28Z -3929307610419788574   
7         user_000001  2009-05-04T13:23:45Z -3929307610419788574   
8         user_000001  2009-05-04T13:19:22Z -3929307610419788574   
9         user_000001  2009-05-04T13:13:38Z -3929307610419788574   
10        user_000001  2009-05-04T13:06:09Z -3929307610419788574   
11        user_000001  2009-05-04T13:00:48Z -3929307610419788574   
12        user_000001  2009-05-04T12:55:34Z -3929307610419788574   
13        user_000001  2009-05-04T12:51:26Z -3929307610419788574   
14        user_000001  2009-05-03T15:48:25Z -4713901082868571884   
15        user_000001  2009-05-03T15:37:56Z -4713901082868571884   
16        user_000001  2009-05-03T15:14:53Z -5244782568945114406   
17        user_000001  2009-05-03T15:10:18Z -8092488190513824180   
18        user_000001  2009-05-03T15:04:31Z  1389540589809226693   
19        user_000001  2009-05-03T14:56:25Z  5825173368809391653   
20        user_000001  2009-05-03T14:50:51Z  1664058600918639843   
21        user_000001  2009-05-03T14:46:29Z -8092488190513824180   
22        user_000001  2009-05-03T14:39:20Z -4946054458147796828   
23        user_000001  2009-05-03T14:34:06Z  7104504505105330621   
24        user_000001  2009-05-03T14:28:02Z -5670205725203648292   
25        user_000001  2009-05-03T14:20:15Z  9100765040592012275   
26        user_000001  2009-05-03T14:07:40Z  9100765040592012275   
27        user_000001  2009-05-03T14:01:15Z  9100765040592012275   
28        user_000001  2009-05-03T13:56:25Z  9100765040592012275   
29        user_000001  2009-05-03T13:47:20Z  9100765040592012275   
30        user_000001  2009-05-03T13:40:46Z  9100765040592012275   
31        user_000001  2009-05-03T13:36:15Z  9100765040592012275   
32        user_000001  2009-05-03T13:32:44Z  -930949355769685053   
33        user_000001  2009-05-03T13:29:11Z  -930949355769685053   
34        user_000001  2009-05-03T13:21:10Z -6250688152594723023   
35        user_000001  2009-05-03T13:13:25Z -6250688152594723023   
36        user_000001  2009-05-03T13:02:08Z -6250688152594723023   
37        user_000001  2009-05-02T15:29:50Z -4713455945721701333   
38        user_000001  2009-05-02T15:24:45Z -4713455945721701333   
39        user_000001  2009-05-02T15:19:46Z -4713455945721701333   
40        user_000001  2009-05-02T15:13:49Z -4713455945721701333   
41        user_000001  2009-05-02T15:08:57Z -4713455945721701333   
42        user_000001  2009-05-02T15:05:06Z -4713455945721701333   
43        user_000001  2009-05-02T15:00:59Z -4713455945721701333   
44        user_000001  2009-05-02T14:56:52Z -4713455945721701333   
45        user_000001  2009-05-02T14:56:52Z -4713455945721701333   
46        user_000001  2009-05-02T14:52:15Z -4713455945721701333   
47        user_000001  2009-05-02T14:50:20Z  2499307466624852415   
48        user_000001  2009-05-02T14:43:46Z  2896184389544170548   
49        user_000001  2009-05-02T14:37:11Z  6910042413922548214   
...               ...                   ...                  ...   
19098812  user_001000  2008-01-28T04:27:23Z  3080918106461434281   
19098813  user_001000  2008-01-28T04:24:11Z  3080918106461434281   
19098814  user_001000  2008-01-28T04:19:16Z  3080918106461434281   
19098815  user_001000  2008-01-28T04:14:45Z  3080918106461434281   
19098816  user_001000  2008-01-28T04:09:30Z  3080918106461434281   
19098817  user_001000  2008-01-28T04:02:10Z  3080918106461434281   
19098818  user_001000  2008-01-28T03:58:51Z  3080918106461434281   
19

_fig 6.1_

---

In [16]:
df.isna().sum()

User_ID        0
Timestamp      0
Artist_ID      0
Artist_Name    0
Track_ID       0
Track_Name     0
dtype: int64

_fig 6.2_

---

> Here we can see that the newly written `ID` columns (_fig 6.1_) have no missing values (_fig 6.2_), it should be noted that the format of the hash has now also changed to a purely numerical format from hexidecimal.

# Data Analysis

> This section will be commited to answering the following 3 questions:  
>
>**A.** Create a list of user IDs, along with the number of distinct songs each user has played.    
>
>**B.** Create a list of the 100 most popular songs (artist and title) in the dataset, with the number of times each was played.
>
>**C.** Say we define a user’s “session” of Last.fm usage to be comprised of one or more songs played by that user, where each song is started within 20 minutes of the previous song’s start time. Create a list of the top 10 longest sessions, with the following information about each session: userid, timestamp of first and last songs in the session, and the list of songs played in the session (in order of play).

## **A.**
_Create a list of user IDs, along with the number of distinct songs each user has played._

> My understanding of this question is for the amount of **distinct** songs, so those that are a unique combination of an `Artist_Name` and `Track_Name`. Such that if I listened to _Nina Simone_ - _Feeling Good_ twice in one day that would only count as 1. Furthermore, if I then listened to _Muse_ - _Feeling Good_ on the same day as the aforementioned songs, this would culminate to 2, as the same song title was produced by 2 different artists.  
With this in mind I have created a new column below (_fig. 7.1_) called `Artist_Track_ID` which is the culmination of the `Artist_ID` and `Track_ID` fields. This is useful because a unique value in this new field won't fall foul of the problem that could arise if multiple artists' had the same `Track_Name`, counting them as none-unique when they are.  
>**P.S.** I know that it says a **list** of user IDs within the task, and I assume would also require a **list** of the number of distinct songs each of the respective users has listened to. However, I have chosen to produce a dataframe and dictionary with the _key_ __:__ _value_ pair of `User_ID` **:** `Unique_Song_Plays` as I belive this to be more suitable.

In [17]:
df['Artist_Track_ID'] = df['Artist_ID'] + df['Track_ID']
df.loc[0]

User_ID                                           user_000001
Timestamp                                2009-05-04T23:08:57Z
Artist_ID                                 -926383222125822968
Artist_Name                                         Deep Dish
Track_ID                                  1237163088474314519
Track_Name         Fuck Me Im Famous (Pacha Ibiza)-09-28-2007
Artist_Track_ID                            310779866348491551
Name: 0, dtype: object

_fig 7.1_

---

In [18]:
Section_A_DF = df.groupby('User_ID').Artist_Track_ID.nunique().sort_values(ascending=True)
Section_A_dict = dict(zip(Section_A_DF.index.tolist(), Section_A_DF.tolist()))
Section_A_dict

{'user_000332': 2,
 'user_000240': 3,
 'user_000686': 5,
 'user_000856': 5,
 'user_000895': 6,
 'user_000566': 6,
 'user_000061': 9,
 'user_000815': 12,
 'user_000538': 16,
 'user_000852': 23,
 'user_000533': 28,
 'user_000231': 41,
 'user_000866': 42,
 'user_000248': 50,
 'user_000618': 51,
 'user_000904': 53,
 'user_000492': 53,
 'user_000982': 54,
 'user_000677': 62,
 'user_000308': 69,
 'user_000563': 70,
 'user_000785': 75,
 'user_000561': 76,
 'user_000608': 76,
 'user_000759': 81,
 'user_000843': 81,
 'user_000455': 85,
 'user_000386': 86,
 'user_000301': 92,
 'user_000272': 100,
 'user_000286': 112,
 'user_000101': 116,
 'user_000846': 119,
 'user_000855': 138,
 'user_000864': 144,
 'user_000985': 146,
 'user_000723': 147,
 'user_000955': 147,
 'user_000597': 149,
 'user_000364': 152,
 'user_000146': 169,
 'user_000548': 170,
 'user_000183': 184,
 'user_000352': 187,
 'user_000335': 192,
 'user_000282': 192,
 'user_000247': 194,
 'user_000140': 197,
 'user_000088': 198,
 'user_

_fig 7.2_

---

In [19]:
Track_Name_Nonunique = len(list(df['Track_Name']))
Track_Name_Unique = len(set(df['Track_Name']))

Artist_Track_Name_Nonunique = len(df['Artist_Track_ID'])
Artist_Track_Name_Unique = len(set(df['Artist_Track_ID']))

print('\t\tTrack_Name\tArtist_Track_Name\nNon-Unique\t{}\t{}\nUnique\t\t{}\t\t{}'.format(Track_Name_Nonunique, \
                                                                                    Artist_Track_Name_Nonunique,\
                                                                                    Track_Name_Unique, \
                                                                                    Artist_Track_Name_Unique))

		Track_Name	Artist_Track_Name
Non-Unique	19098850	19098850
Unique		1083480		1498151


_fig 7.3_

---

>Above in _fig 7.2_ if my solution to **Section A** with the dictionary format `User_ID`:`Unique_Song_Plays`.   
In _fig 7.3_ I show what I discussed at the start of this section, as both `Track_Name` and `Artist_Track_ID` have the same values whilst contained in lists, allowing for duplicate values.   
However, when both are held in sets instead, which remove duplicates implicitly, you can see that the number of fields in `Artist_Track_ID` > `Track_Name` as each record is a combination of `Track_Name` and `Artist_Name` fields. This allows for instances where multiple bands will have the same `Track_Name`, but will still count them as distinct.

## **B.**
_Create a list of the 100 most popular songs (artist and title) in the dataset, with the number of times each was played._

In [20]:
Section_A_DF = df.groupby(['Artist_Name', 'Track_Name']).Artist_Track_ID.count().nlargest(100)
Section_A_DF

Artist_Name            Track_Name                                     
The Postal Service     Such Great Heights                                 3991
Boy Division           Love Will Tear Us Apart                            3651
Radiohead              Karma Police                                       3533
Death Cab For Cutie    Soul Meets Body                                    3479
Muse                   Supermassive Black Hole                            3463
The Knife              Heartbeats                                         3155
Arcade Fire            Rebellion (Lies)                                   3047
Muse                   Starlight                                          3040
Britney Spears         Gimme More                                         3002
The Killers            When You Were Young                                2997
Interpol               Evil                                               2962
Kanye West             Love Lockdown                        

_fig 8.1_

In [21]:
df.loc[df['Track_Name'] == 'Map Of The Problematique']

User_ID             Timestamp            Artist_ID Artist_Name  \
17053     user_000002  2009-04-07T17:41:51Z  7940227560495178971        Muse   
17554     user_000002  2009-03-11T14:03:43Z  7940227560495178971        Muse   
17574     user_000002  2009-03-11T12:46:19Z  7940227560495178971        Muse   
17584     user_000002  2009-03-11T12:04:12Z  7940227560495178971        Muse   
160008    user_000006  2006-07-12T22:02:59Z  7940227560495178971        Muse   
164501    user_000008  2009-04-18T21:07:31Z  7940227560495178971        Muse   
175299    user_000008  2009-02-25T17:13:55Z  7940227560495178971        Muse   
175313    user_000008  2009-02-25T16:14:20Z  7940227560495178971        Muse   
175316    user_000008  2009-02-25T16:02:20Z  7940227560495178971        Muse   
175321    user_000008  2009-02-25T15:44:00Z  7940227560495178971        Muse   
175335    user_000008  2009-02-25T14:44:26Z  7940227560495178971        Muse   
175349    user_000008  2009-02-25T13:44:50Z  7940227560495178971        Muse   
175356    user_000008  2009-02-25T13:12:16Z  7940227560495178971        Muse   
175369    user_000008  2009-02-25T12:17:41Z  7940227560495178971        Muse   
175383    user_000008  2009-02-25T11:18:06Z  7940227560495178971        Muse   
175397    user_000008  2009-02-25T10:18:31Z  7940227560495178971        Muse   
175411    user_000008  2009-02-25T09:18:55Z  7940227560495178971        Muse   
175428    user_000008  2009-02-25T04:29:07Z  7940227560495178971        Muse   
175445    user_000008  2009-02-25T02:29:56Z  7940227560495178971        Muse   
175451    user_000008  2009-02-24T23:11:43Z  7940227560495178971        Muse   
178858    user_000008  2009-02-14T03:48:54Z  7940227560495178971        Muse   
179089    user_000008  2009-02-14T01:49:42Z  7940227560495178971        Muse   
192127    user_000008  2009-01-13T22:01:24Z  7940227560495178971        Muse   
192141    user_000008  2009-01-13T21:01:48Z  7940227560495178971        Muse   
192155    user_000008  2009-01-13T20:02:14Z  7940227560495178971        Muse   
192169    user_000008  2009-01-13T19:02:39Z  7940227560495178971        Muse   
192183    user_000008  2009-01-13T18:03:04Z  7940227560495178971        Muse   
192198    user_000008  2009-01-13T16:03:53Z  7940227560495178971        Muse   
192212    user_000008  2009-01-13T15:04:17Z  7940227560495178971        Muse   
192226    user_000008  2009-01-13T14:04:42Z  7940227560495178971        Muse   
192240    user_000008  2009-01-13T13:05:06Z  7940227560495178971        Muse   
192254    user_000008  2009-01-13T12:05:31Z  7940227560495178971        Muse   
192268    user_000008  2009-01-13T11:05:55Z  7940227560495178971        Muse   
192290    user_000008  2009-01-13T09:24:53Z  7940227560495178971        Muse   
192304    user_000008  2009-01-13T08:25:17Z  7940227560495178971        Muse   
192318    user_000008  2009-01-13T07:25:41Z  7940227560495178971        Muse   
192332    user_000008  2009-01-13T06:26:05Z  7940227560495178971        Muse   
192346    user_000008  2009-01-13T05:26:30Z  7940227560495178971        Muse   
192360    user_000008  2009-01-13T04:26:54Z  7940227560495178971        Muse   
192374    user_000008  2009-01-13T03:27:18Z  7940227560495178971        Muse   
192388    user_000008  2009-01-13T02:27:43Z  7940227560495178971        Muse   
192402    user_000008  2009-01-13T01:28:07Z  7940227560495178971        Muse   
192409    user_000008  2009-01-13T00:51:17Z  7940227560495178971        Muse   
192423    user_000008  2009-01-12T23:51:41Z  7940227560495178971        Muse   
192426    user_000008  2009-01-12T23:34:01Z  7940227560495178971        Muse   
215136    user_000011  2007-12-23T12:44:33Z  7940227560495178971        Muse   
313277    user_000013  2007-01-25T18:12:15Z  7940227560495178971        Muse   
313281    user_000013  2007-01-25T17:58:14Z  7940227560495178971        Muse   
313353    user_000013  2007-01-21T22:27:23Z  7940227560495178971        Muse   
313357    user

_fig 8.2_

---

> _fig 8.1_ shows the list of the top 100 most played songs, along with the number of times they were played and the songs artist.  
It should be noted that there are 3 songs in this list (_'Map Of The Problematique'_, _'Robocop'_, _'See You In My Nightmares'_) that look like they don't have a `Artist_Name` attached to them. They do have this information available in the dataframe (for example 'Map Of The Problematique' in _fig 8.2_) and is instead what I believe is a result of a rendering bug.

## **C.**
_Say we define a user’s “session” of Last.fm usage to be comprised of one or more songs played by that user, where each song is started within 20 minutes of the previous song’s start time. Create a list of the top 10 longest sessions, with the following information about each session: userid, timestamp of first and last songs in the session, and the list of songs played in the session (in order of play)._

In [31]:
df['Timestamp'] = pd.to_datetime(df['Timestamp'])
df['Track_Gap'] = ''

_fig 9.1_

---

In [119]:
df.dtypes

User_ID                         object
Timestamp          datetime64[ns, UTC]
Artist_ID                        int64
Artist_Name                     object
Track_ID                         int64
Track_Name                      object
Artist_Track_ID                  int64
Timestamp_dttm     datetime64[ns, UTC]
Track_Gap                       object
dtype: object

_fig 9.2_

---

In [66]:
ID = 'user_000001'
%timeit df.loc[df['User_ID'] == ID, ['Track_Gap']] = (df['Timestamp_dttm']  - df['Timestamp_dttm'].shift(-1)).dt.total_seconds()

2.69 s ± 137 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


_fig 9.3_

---

In [758]:
user_length = len(df.loc[df['User_ID'] == 'user_000001'])
df_length = len(df)
estimated_time = (round((df_length/user_length) * 2.55)/60) # Rounding down from minimum average calculated fig 9.3

print('dataframe length:\t{}\nuser_000001 length:\t{}\n\nestimated time:\t\t{} minutes'.format(user_length, df_length, estimated_time))

dataframe length:	16685
user_000001 length:	19098850

estimated time:		48.65 minutes


_fig 9.4_

---

In [705]:
min_date = '2008-01-01T00:00:00Z'
max_date = '2009-01-01T00:00:00Z'
df_subset = df.loc[(df['Timestamp'] > min_date) & (df['Timestamp'] < max_date)].sort_values(by=['User_ID', 'Timestamp'], ascending = True)

_fig 9.5_

---

In [706]:
track_gap_list = []

for ID in df_subset['User_ID'].unique():
    
    df_user_subset = df_subset.loc[df_subset['User_ID'] == ID, ['Timestamp','Track_Gap']]
    df_user_subset['Track_Gap'] = (df_user_subset['Timestamp'].shift(-1) - df_user_subset['Timestamp']).dt.total_seconds()

    df_user_subset.loc[df_user_subset['Track_Gap'] > 1200, 'Track_Gap'] = -99

    songs = df_user_subset['Track_Gap'].tolist()

    track_gap_list.extend(songs)

df_subset['Track_Gap'] = track_gap_list
df_subset["Session_Length"] = df_subset["Track_Gap"].groupby((df_subset["Track_Gap"] == -99).cumsum()).cumcount()
df_subset.reset_index()

user_000001 : 7917
user_000002 : 12389
user_000003 : 6026
user_000004 : 6865
user_000005 : 3589
user_000006 : 5271
user_000007 : 359
user_000008 : 3986
user_000009 : 1898
user_000010 : 2879
user_000011 : 904
user_000012 : 20300
user_000014 : 1
user_000015 : 11321
user_000016 : 4308
user_000017 : 9708
user_000019 : 1645
user_000020 : 615
user_000021 : 14636
user_000022 : 15635
user_000023 : 27027
user_000024 : 3511
user_000025 : 8160
user_000026 : 15380
user_000027 : 2131
user_000028 : 164
user_000029 : 10001
user_000030 : 11610
user_000031 : 8990
user_000032 : 5095
user_000033 : 12456
user_000034 : 3544
user_000035 : 7138
user_000036 : 2669
user_000037 : 1711
user_000038 : 9048
user_000040 : 6605
user_000041 : 3473
user_000042 : 50
user_000043 : 2752
user_000044 : 6126
user_000046 : 2270
user_000047 : 154
user_000049 : 22
user_000050 : 1205
user_000051 : 8288
user_000052 : 5532
user_000053 : 6276
user_000054 : 24124
user_000055 : 3322
user_000056 : 2904
user_000058 : 37
user_000059 : 1

user_000515 : 37140
user_000516 : 5927
user_000517 : 15534
user_000518 : 5793
user_000520 : 3201
user_000522 : 6400
user_000523 : 2446
user_000524 : 9580
user_000526 : 1
user_000527 : 812
user_000528 : 89
user_000529 : 70323
user_000531 : 1254
user_000532 : 1810
user_000534 : 712
user_000535 : 2144
user_000536 : 36373
user_000537 : 3693
user_000539 : 211
user_000540 : 17223
user_000541 : 2
user_000542 : 4014
user_000543 : 21871
user_000544 : 60129
user_000545 : 1
user_000546 : 5395
user_000549 : 11877
user_000550 : 2717
user_000551 : 4067
user_000552 : 4997
user_000553 : 6509
user_000554 : 15670
user_000557 : 205
user_000558 : 3205
user_000559 : 123
user_000560 : 249
user_000561 : 250
user_000564 : 8675
user_000565 : 15524
user_000566 : 2
user_000567 : 1
user_000568 : 39287
user_000569 : 12662
user_000570 : 14525
user_000571 : 11887
user_000572 : 6920
user_000573 : 10027
user_000574 : 7012
user_000575 : 185
user_000576 : 2343
user_000577 : 26310
user_000578 : 1317
user_000579 : 3650
us

index      User_ID                 Timestamp            Artist_ID  \
0           12043  user_000001 2008-01-01 06:57:28+00:00  7379093288854802642   
1           12042  user_000001 2008-01-06 13:41:36+00:00  7379093288854802642   
2           12041  user_000001 2008-01-06 13:59:14+00:00  -864670504031739368   
3           12040  user_000001 2008-01-06 14:49:30+00:00  -864670504031739368   
4           12039  user_000001 2008-01-06 14:52:42+00:00  -864670504031739368   
5           12038  user_000001 2008-01-07 12:40:17+00:00  -864670504031739368   
6           12037  user_000001 2008-01-07 12:43:01+00:00  -864670504031739368   
7           12036  user_000001 2008-01-07 12:45:52+00:00  -864670504031739368   
8           12035  user_000001 2008-01-07 12:48:30+00:00  -864670504031739368   
9           12034  user_000001 2008-01-07 12:51:48+00:00  -864670504031739368   
10          12033  user_000001 2008-01-07 12:56:33+00:00  -864670504031739368   
11          12032  user_000001 2008-01-07 12:58:50+00:00  -864670504031739368   
12          12031  user_000001 2008-01-07 13:01:03+00:00  -864670504031739368   
13          12030  user_000001 2008-01-08 12:52:03+00:00 -1441634570900672991   
14          12029  user_000001 2008-01-08 12:55:38+00:00 -1441634570900672991   
15          12028  user_000001 2008-01-08 13:03:51+00:00 -1441634570900672991   
16          12027  user_000001 2008-01-08 13:10:58+00:00 -1441634570900672991   
17          12026  user_000001 2008-01-08 13:22:00+00:00 -1441634570900672991   
18          12025  user_000001 2008-01-08 13:27:53+00:00 -1441634570900672991   
19          12024  user_000001 2008-01-08 13:33:51+00:00 -1441634570900672991   
20          12023  user_000001 2008-01-09 04:18:55+00:00 -1441634570900672991   
21          12022  user_000001 2008-01-09 04:29:58+00:00 -1441634570900672991   
22          12021  user_000001 2008-01-09 04:35:50+00:00 -1441634570900672991   
23          12020  user_000001 2008-01-09 04:41:49+00:00 -1441634570900672991   
24          12019  user_000001 2008-01-09 04:48:01+00:00 -1441634570900672991   
25          12018  user_000001 2008-01-09 04:54:35+00:00 -1441634570900672991   
26          12017  user_000001 2008-01-09 04:59:51+00:00 -1441634570900672991   
27          12016  user_000001 2008-01-09 05:30:07+00:00 -1441634570900672991   
28          12015  user_000001 2008-01-09 05:37:10+00:00  -864670504031739368   
29          12014  user_000001 2008-01-09 05:40:27+00:00  -864670504031739368   
30          12013  user_000001 2008-01-09 05:44:13+00:00 -1441634570900672991   
31          12012  user_000001 2008-01-09 05:47:48+00:00 -1441634570900672991   
32          12011  user_000001 2008-01-09 05:56:01+00:00 -1441634570900672991   
33          12010  user_000001 2008-01-09 06:03:08+00:00 -1441634570900672991   
34          12009  user_000001 2008-01-09 06:28:57+00:00 -1441634570900672991   
35          12008  user_000001 2008-01-09 06:34:49+00:00 -1441634570900672991   
36          12007  user_000001 2008-01-09 06:40:48+00:00 -1441634570900672991   
37          12006  user_000001 2008-01-09 07:20:05+00:00 -1441634570900672991   
38          12005  user_000001 2008-01-09 07:29:45+00:00 -1441634570900672991   
39          12004  user_000001 2008-01-09 07:35:00+00:00 -1441634570900672991   
40          12003  user_000001 2008-01-09 07:41:50+00:00 -1441634570900672991   
41          12002  user_000001 2008-01-09 07:53:04+00:00 -1441634570900672991   
42          12001  user_000001 2008-01-09 07:56:39+00:00 -1441634570900672991   
43          12000  user_000001 2008-01-09 08:04:52+00:00 -1441634570900672991   
44          11999  user_000001 2008-01-09 08:11:59+00:00 -1441634570900672991   
45          11998  user_000001 2008-01-09 08:23:01+00:00 -1441634570900672991   
46          11997  user_000001 2008-01-09 08:28:53+00:00 -1441634570900672991   
47          11996  user_000001 2008-01-09 08:34:52+00:00 -1441634570900672991   
48          11995  user_000001 2008-01-09 0

_fig 9.6_

---

In [707]:
df_subset_10 = df_subset.groupby(['User_ID'])['Session_Length'].max().nlargest(10).reset_index()

user_id_list = df_subset_10['User_ID'].tolist()
session_list = df_subset_10['Session_Length'].tolist()

top_session_df = pd.DataFrame(columns=['User_ID', 'Track_Name', 'Start_Timestamp', 'End_Timestamp'])

_fig 9.7_

---

In [745]:
pd.options.mode.chained_assignment = None


for k, user in enumerate(user_id_list):
    
    session_length = session_list[k]
    
    end_label = df_subset.loc[(df_subset['User_ID'] == user) & (df_subset['Session_Length'] == session_length)].index.array[0]
    end = df_subset.index.get_loc(end_label) + 1
    start = end - session_length

    working_set = df_subset.iloc[start:end]
    working_set.drop(columns=['Artist_ID', 'Artist_Name', 'Track_ID', 'Artist_Track_ID', 'Timestamp_dttm', 'Track_Gap', 'Session_Length'], inplace=True)
    working_set['Start_Timestamp'] = working_set['Timestamp'].min()
    working_set['End_Timestamp'] = working_set['Timestamp'].max()
    
    working_set.drop(columns=['Timestamp'], inplace=True)
        
    top_session_df = top_session_df.append(working_set, ignore_index=True, sort=True)
    

top_session_df = top_session_df[['User_ID', 'Track_Name', 'Start_Timestamp', 'End_Timestamp']]

top_session_df

_fig 9.8_

---

> My process for apporaching this question was: 
>
> > 1. Calculate the `Track_Gap` to identify the sessions
> > 2. Identify the longest sessions
> > 3. Return all the preceeding records by utilising the **max** record as a base point and `Session_Length` to find the first record.
>
> Initially this required me to change the type of data held in `df['Timestamp']` from string to timestamp, achieved in _fig 9.1_; the results being shown in _fig 9.2_.  
Once I started to apply **step 1.**, it quickly became apparent that this would take a large amount of time to run through the ~ 20 million records present in the dataset. With my estimate from _fig 9.3 & 9.4_ putting just **step 1.** at 48.7 minutes.  Due to this I decided to filter the dataset I would work on, knowing any logic I used here would work on the rest of the data, only taking longer to run.   
I could have gone down the root of implementing more powerful processing methods such as the writing this in Cython or using the `multiprocessing` module; but based on the task I was completing I am content with my choice.
>
> From here onwards the dataframe is composed of information from the original dataset between `01-01-2008 00:00:00` to `01-01-2009 00:00:00` in `df_subset` _fgi 9.5_.  
>
> In _fig 9.6_ I apply **step 1.** iteratively to all `User_ID`s in `df_subset`, marking all breaks in any session (> 1200 seconds) as `= -99`. This allowed me to easily identify the breaks in sessions in **step 2.**.   
> _fig 9.7_ shows **step 2.**, the selection of the top 10 longest sessions in to `df_subset_10` from `df_subset` and seperation of the series `User_ID` and `Session_Length` in to their respective list objects, `user_id_list` and `session_list`.  
> Finally, _fig 9.8_ shows **step 3.**, the iterative calculation of each `User_ID`s dataframe, in order from longest to shortest session length from the top 10. Each of these is appended to the `top_session_df` a long with the calculated fields `Start_Timestamp` and `End_Timestamp`.